File: 0034-find-first-and-last-position-of-element-in-sorted-array/34.go
package main

import "math"

func first(nums []int, target int) int {

	n := len(nums)

	if n == 0 {
		return -1
	}

	L, R := 0, n-1

	for L != R {
		m := int(math.Floor(float64(L) + float64((R-L))/2))

		if nums[m] >= target {
			R = m
		} else {
			L = m + 1
		}

	}

	if nums[L] == target {
		return L
	}

	return -1

}

func last(nums []int, target int) int {
	n := len(nums)

	if n == 0 {
		return -1
	}

	L, R := 0, n-1

	for L != R {
		m := int(math.Ceil(float64(L) + float64((R-L))/2))

		if nums[m] <= target {
			L = m
		} else {
			R = m - 1
		}

	}

	if nums[L] == target {
		return L
	}

	return -1
}

func searchRange(nums []int, target int) []int {

	a := first(nums, target)
	b := last(nums, target)

	return []int{a, b}

}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/34.cpp
#include <math.h>
#include <vector>

using namespace std;

class Solution {
public:
  int first(vector<int> &nums, int target) {

    int n = nums.size();

    if (n == 0) {
      return -1;
    }

    int L = 0, R = n - 1;

    while (L != R) {
      int m = floor(L + static_cast<float>(R - L) / 2);

      if (nums[m] >= target) {

        R = m;

      } else {

        L = m + 1;
      }
    }

    if (nums[L] == target) {
      return L;
    }

    return -1;
  }

  int last(vector<int> &nums, int target) {
    int n = nums.size();

    if (n == 0) {
      return -1;
    }

    int L = 0, R = n - 1;

    while (L != R) {
      int m = ceil(L + static_cast<float>(R - L) / 2);
      if (nums[m] <= target) {
        L = m;
      } else {
        R = m - 1;
      }
    }

    if (nums[L] == target) {
      return L;
    }

    return -1;
  }
  vector<int> searchRange(vector<int> &nums, int target) {

    int a = first(nums, target);
    int b = last(nums, target);

    return {a, b};
  }
};